Degeneracy Hunter

adowling2 opened the pull request:
## Fixes


## Summary/Motivation:
- Adds a collection of diagnostic tools based on https://www.sciencedirect.com/science/article/pii/B9780444635785501304

## Changes proposed in this PR:
- Add Degeneracy Hunter package to IDAES core

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

## Outstanding Tasks for this PR
- [x] Add IDAES copywrite text
- [x] Remove functions from Robby.
- [x] Migrate improvements back into IDAES core
   - [x] `check_residuals` --> `model_statistics.large_residuals_set`
   - [x] `check_variable_bounds` --> `check_variable_bounds`
- ~~Use IDAES logger to manage output to the console (is this needed)~~ waiting for official release
- [x] Create unit tests
- [x] Improve usability
   - [x] Warn user if attempting to call `check_rank_equality_constraints` on a model without equality constraints
   - [x] Add more verbose output to `check_rank_equality_constraints`
   - [x] Update `find_irreducible_degenerate_sets` to call `find_candidate_equations` if there are no candidate equations
   - [x] Allow user to specify MILP solver

## Tasks for Future PRs
- Add notebook with examples: https://github.com/IDAES/examples-pse/pull/21
- Create documentation page
- Revisit how to extend degeneracy hunter to active bounds/inequality constraints
- Add support for fixed variables
   - Prepare example with fixed variables. Test what happens.
   - If needed, remove columns from the Jacobian that correspond to fixed variable
- How should this interface with the IDAES model scaling features?
- Add tests to pynumero to ensure the required API calls do not change

adowling2 commented:
@Robbybp I just migrated Degeneracy Hunter from workspace to idaes-pse. I'm using two functions you wrote as a workaround for pynumero: `get_con_eq_idx_map` and `get_eq_con_list`. What are your thoughts on adding this functionality to pynumero?

Robbybp commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,577 @@
+    eq_names = list(temp[mask])
+    return {name: i for i, name in enumerate(eq_names)}
+
+def get_eq_con_list(interface, constraint_names):
+    mask = interface._con_full_eq_mask
+    temp = np.array(constraint_names)
+    eq_names = list(temp[mask])
+    return eq_names
```
In my code, I now have this implemented as

```python
def get_eq_con_list(nlp):
    constraints = nlp.get_pyomo_constraints()
    mask = nlp._con_full_eq_mask
    return [con for con, bit in zip(constraints, mask) if bit]
```

Robbybp commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,577 @@
+This code is from Robby Parker.
+'''
+
+def get_con_eq_idx_map(interface, constraint_names):
+    # This is super janky. TODO: Add functionality
+    # to PyNumero.
+    # constraint_names is an argument because generating
+    # constraint_names list from PyNumero takes forever.
+    # TODO: Find way to make AslNLP::constraint_names()
+    # not ridiculously slow.
+    mask = interface._con_full_eq_mask
+    temp = np.array(constraint_names)
+    eq_names = list(temp[mask])
+    return {name: i for i, name in enumerate(eq_names)}
```
I believe a better implementation would be

```python
def get_con_eq_idx_map(nlp):
    eq_cons = get_eq_con_list(nlp) # See my implementation of function below
    return ComponentMap((con, i) for i, con in enumerate(eq_cons))
```
This avoids needing to generate constraint names, which is slow.

adowling2 commented:
@Robbybp Thank you for the updates!! Are these functions already somewhere else in Pyomo or IDAES? Or should I keep them here for now?

Robbybp commented:
> @Robbybp I just migrated Degeneracy Hunter from workspace to idaes-pse. I'm using two functions you wrote as a workaround for pynumero: `get_con_eq_idx_map` and `get_eq_con_list`. What are your thoughts on adding this functionality to pynumero?

I left some comments below with re-implementations of these functions that I think should still fit your needs. These re-implementations are more general. It looks like this functionality exists in PyNumero in the AmplNLP class, (`eq_constraint_names`, `constraint_idx`), but not in PyomoNLP, where we have `get_pyomo_constraints`, `constraint_names`, and `get_constraint_indices`, but not the corresponding functions for equality constraints. This should definitely be added - I'll look into doing so shortly. 

Do you need to access indices from names or the Pyomo objects?

Robbybp commented:
> Are these functions already somewhere else in Pyomo or IDAES? Or should I keep them here for now?

I just opened PR https://github.com/Pyomo/pyomo/pull/1791 that adds this functionality to the `PyomoNLP` class in PyNumero.

adowling2 commented:
@Robbybp Hopefully this is a quick yes/no:

**Question 1**

Right now, Degeneracy Hunter uses
```
def get_eq_con_list(interface, constraint_names):
    mask = interface._con_full_eq_mask
    temp = np.array(constraint_names)
    eq_names = list(temp[mask])
    return eq_names
```

should this get replaced with `equality_constraint_names` in `PyomoNLP`?

**Question 2**

Degeneracy Hunter uses
```
def get_con_eq_idx_map(interface, constraint_names):
    # constraint_names is an argument because generating
    # constraint_names list from PyNumero takes forever.
    # TODO: Find way to make AslNLP::constraint_names()
    # not ridiculously slow.
    mask = interface._con_full_eq_mask
    temp = np.array(constraint_names)
    eq_names = list(temp[mask])
    return {name: i for i, name in enumerate(eq_names)}
```

should this get replaced with `get_pyomo_equality_constraints` in `PyomoNLP`?

**Question 3**

The big difference regarding the functions in Q2 that `get_con_eq_idx_map` returns a dictionary where the constraint name is the index whereas `get_pyomo_equality_constraints` returns a list? Is the list returned by `get_pyomo_equality_constraints` in the same order as the list returned by `equality_constraint_names`?

I'm looking at the code in this recently merged PR. https://github.com/Pyomo/pyomo/pull/1791

adowling2 commented:
@lbianchi-lbl I need to create unit tests for this. Can I just copy https://github.com/IDAES/idaes-pse/blob/main/idaes/core/util/tests/test_model_statistics.py in the same folder, rename to `test_model_diagnostics.py` and update accordingly? Do I need to do anything else special?

Robbybp commented:
@adowling2
1. If you want to get a list of Pyomo equality constraints, you should call `PyomoNLP.get_pyomo_equality_constraints`. If you want the names, you can use `equality_constraint_names`, although this is very slow and not recommended.
2. No. It should be replaced by `PyomoNLP.get_equality_constraint_indices`. Of course in this case you need to use the Pyomo objects as keys, which is probably fine.
3. Right. The dict and list are essentially inversions of each other. Those lists are in the same order.

ksbeattie commented:
@adowling2, is this still on-track for being merged in time for cutting a 1.9 release candidate next week, on the 11th?

adowling2 commented:
@ksbeattie I think so.

adowling2 commented:
Quick update (mainly a note for myself):
* Incorporated code from Robby

Next steps:
- [ ] Write unit tests. Mainly look at check_residuals() function
- [ ] Open issue for DH future improvements
    - Write documentation
    - Reorganize class structure
    - Add cbc to the testing harness
    - Add unit tests for degeneracy hunter (needs MILP solver)
- [ ] Check desktop for missing push - just minor fix

adowling2 commented:
This is ready for review! @jghouse88 @andrewlee94 @Robbybp @ksbeattie

The added tests pass on my local machine. I have one outstanding minor change on my desktop (at work) I need to commit Monday AM. It does better error handling for one case in the example.

My hope to for this to make it into the IDAES 1.9 release. I'll then use the feature and test it in my elective this semester. During the summer/fall, I plan to work on several enhancements (#204), including proper documentation. Right now this is a not advertised feature, although a "power user" can figure it out from my example IDAES/examples-pse#21.

@lbianchi-lbl Do I need to do anything special to trigger the tests?

adowling2 commented:
This is ready to review. I merged in the minor changes that were on my desktop. (If you read the messages above, that was the last thing left.)

andrewlee94 commented on idaes/core/util/tests/test_model_diagnostics.py:
```diff
@@ -0,0 +1,249 @@
+    
+    assert n_rank_deficient == 1
+    
+    # TODO: Add MILP solver to idaes get-extensions and add more tests
```
I agree that this is needed. It would also be nice to test the printed/logged output of some of the functions. Pytest captures output to the logger, stdout and stderr which can be used to check for expected output strings.

adowling2 replied:
This is a great idea. I added it to the TODO list.

andrewlee94 reviewed (commented):
A few general comments - I will eave it to the experts to look over the actual methods in detail.

The biggest thing I note is the lack of documentation - we should at a minimum add a placeholder page to the user docs saying that this is a WIP (at the very least to make sure we don't forget). I also noted that documentation of the expected arguments for most of the methods was incomplete - we should make sure to mention all the arguments in the doc strings.

Given the limited timeframe before the release, we probably can't address all of these comments before we need to merge this, but we should make sure we add these to an issue of the next steps required.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+                
+        # Initialize solver
+        if solver is None:
+            self.solver = SolverFactory('gurobi')
```
Yes, once we add cbc to the idaes binaries.

adowling2 replied:
I added configurable options to the TODO list.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+            tol : residual threshold for inclusion in ComponentSet (default = 1e-5)
+            relative : Boolean, use relative tolerance (default = False)
+            skip_lb: Boolean to skip lower bound (default = False)
+            skip_ub: Boolean to skip upper bound (default = False)
```
Using the IDAES logger is on my to-do list for the next release. I'll see if I can get it working before Thursday.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+            self.candidate_eqns = None
+        
+        else:
+            self.jac_eq = block_or_jac
```
I added a NotImplementedError exception. I should really refactor the code, make an example, test it, document it, ...

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+
+# This library is already used in Pyomo
+import operator
+
```
I fixed this and added PEP8 to the TODO issue.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+
+__author__ = "Alexander Dowling"
+
+from pyomo.environ import *
```
Fixed, thank you.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+    
+        Arguments:
+            block_or_jac: Pyomo model or Jacobian
+        
```
Fixed, thank you.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+        self.min_nonzero_nu = 1E-5
+        
+        
+    def check_residuals(self,tol=1e-5, print_level=2, sort=True):
```
This is on the TODO list. I agree.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+    def check_rank_equality_constraints(self,tol=1E-6):
+        """
+        Method to check the rank of the Jacobian of the equality constraints
+        
```
Fixed.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+    def svd_analysis(self,n_smallest_sv = 10):
+        '''
+        Perform SVD analysis of the constraint Jacobian
+        '''
```
Fixed.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+    
+    def find_candidate_equations(self,verbose=True,tee=False):
+        '''
+        Solve MILP to find a degenerate set and candidate equations
```
Fixed.

adowling2 commented:
@andrewlee94 I addressed all of your comments and added outstanding tasks to #204.

@Robbybp @jghouse88 @jalving @jsiirola @xiangao1 @blnicho Any reviews you can provide before tomorrow afternoon are appreciated. You can test the code by running the example notebook here: https://github.com/IDAES/examples-pse/pull/21

blnicho commented on idaes/core/util/model_statistics.py:
```diff
@@ -1284,13 +1296,45 @@ def large_residuals_set(block, tol=1e-5):
+        r = 0.0 # residual
+        
+        # check the lower bound
+        # skip if inequality constraint
```
This comment is a bit misleading because the logic as-written will still compute residuals for violated bounds on inequality constraints. Maybe just say "skip if no lower bound set"

adowling2 replied:
Good suggestion.

blnicho commented on idaes/core/util/model_statistics.py:
```diff
@@ -1284,13 +1296,45 @@ def large_residuals_set(block, tol=1e-5):
+            r = r_temp
+    
+        # check the upper bound
+        # skip if inequality constraint
```
Same note about the misleading comment

adowling2 replied:
Fixed.

blnicho commented on idaes/core/util/model_statistics.py:
```diff
@@ -1270,7 +1282,7 @@ def degrees_of_freedom(block):
 
 
-def large_residuals_set(block, tol=1e-5):
+def large_residuals_set(block, tol=1e-5, return_residual_values=False):
```
The `return_residual_values` argument is not documented. Also, I think it's critical to document that the return type for the function will change depending on the value provided to that argument

adowling2 replied:
Great point.

blnicho commented on idaes/core/util/tests/test_model_diagnostics.py:
```diff
@@ -0,0 +1,249 @@
+    m = problem1()
+    
+    # Specify Ipopt as the solver
+    opt = pyo.SolverFactory('ipopt')
```
Could you add skip conditions to these tests if ipopt is not available? See here for an example: https://github.com/IDAES/idaes-pse/blob/main/idaes/tests/test_logger.py#L95

andrewlee94 replied:
Personally, I have been pushing to drop the `skipifs`. Seeing as we have IPOPT easily available as part of get_extensions, I see no reason why we should not expect anyone running the tests to have IPOPT available.

blnicho replied:
I think we should keep the skipifs until we get the Mac ipopt binary added to our extensions

adowling2 replied:
I added it, but agree we should remove it once we have macOS binaries.

blnicho commented on idaes/core/util/tests/test_model_diagnostics.py:
```diff
@@ -0,0 +1,249 @@
+    # Check second constraint
+    assert initial_point_constraint_names[1] == 'con3'
+    
+    # Specifying an iteration limit of 0 allows us to inspect the initial point
```
Looks like this comment needs to be updated or removed

adowling2 replied:
Good catch.

blnicho commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,611 @@
+            
+        else:
+            # TODO: Make this a custom exception following IDAES standards
+            assert type(solver) is SolverFactory, "Argument solver should be type SolverFactory" 
```
I don't think the class type is correct for the `solver` argument. I think the `SolverFactory` returns a solver object which isn't a subclass of `SolverFactory` so this assert statement will never be true. @jsiirola is there a solver base class that could be checked for here?

Robbybp replied:
I have used `SystemCallSolver` in the past, but at some point changed to the opinion that a solver is anything that implements a `solve` method (probably should have an `options` data structure as well).

adowling2 replied:
I commented this out to be safe.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+
+class DegeneracyHunter():
+
+    def __init__(self,block_or_jac,solver=None):
```
Fixes.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+        
+        '''
+        
+        if type(block_or_jac) is ConcreteModel:
```
Good suggestion.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+        self.s = None
+        
+        # Set constants for MILPs
+        self.max_nu = 1E5
+        self.min_nonzero_nu = 1E-5
```
I added it to the TODO list (issue).

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+        self.min_nonzero_nu = 1E-5
+        
+        
+    def check_residuals(self,tol=1e-5, print_level=2, sort=True):
```
The tolerances set in __init__.py are used for the degeneracy hunting MILP. These are for setting the residuals. This is actually the one spot it makes sense to be function arguments. This function will print out all of the residuals smaller than `tol`. I find it helpful to call this multiple times in a notebook with different values for `tol`.

But I agree, I should use a ConfigDict for most things. It is on the TODO list!

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+        else:
+            return large_residuals_set(self.block, tol, False)
+        
+        if print_level > 0:
```
Good catch.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+                    residual_values = dict(sorted(residual_values.items(), key=operator.itemgetter(1),reverse=True))
+        
+                for i, (c,r) in enumerate(residual_values.items()):
+                    if print_level == 0:
+                        # Basic print statement. count, constraint, residual
+                        print(i,"\t",c,"\t",r)
```
Good catch. Also, using the logger is on the TODO list.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+        m_dh = ConcreteModel()
+
+        # Create index for constraints
+        m_dh.C = Set(initialize=[i for i in range(n_eq)])
+
+        m_dh.V = Set(initialize=[i for i in range(n_var)])
```
Good catch.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,558 @@
+from pyomo.core.expr.visitor import identify_variables
+from pyomo.contrib.pynumero.interfaces.pyomo_nlp import PyomoNLP
+import numpy as np
+from scipy.sparse.linalg import svds
+from scipy.sparse import issparse, find
```
Done, although someone who knows how `setup.py` should check this.

adowling2 commented on idaes/core/util/model_statistics.py:
```diff
@@ -1284,13 +1296,45 @@ def large_residuals_set(block, tol=1e-5):
+        else:
+            r_temp = value(c.lower - c.body())
+        # update the residual
+        if c.active and r_temp > r:
```
Good catch.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,611 @@
+
+        ids_milp.nu[c].unfix()
+
+        if results.solver.Status == SolverStatus.ok and results.solver.termination_condition == TerminationCondition.optimal:
```
I like it.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,611 @@
+        
+        results = solver.solve(candidates_milp, tee=tee)
+        
+        if results.solver.Status == SolverStatus.ok and results.solver.termination_condition == TerminationCondition.optimal:
```
Done.

adowling2 commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,611 @@
+        """
+        
+        # If there are no candidate equations, find them!
+        if self.candidate_eqns is None or not self.candidate_eqns:
```
Done.

adowling2 commented on idaes/core/util/model_statistics.py:
```diff
@@ -1284,13 +1296,45 @@ def large_residuals_set(block, tol=1e-5):
+        else:
+            r_temp = value(c.lower - c.body())
+        # update the residual
+        if c.active and r_temp > r:
```
Good catch!

adowling2 commented on idaes/core/util/model_statistics.py:
```diff
@@ -1284,13 +1296,45 @@ def large_residuals_set(block, tol=1e-5):
+            r_temp = value(c.body() - c.upper)
+
+        # update the residual
+        if c.active and r_temp > r:
```
Good catch!

adowling2 commented:
@blnicho @jsiirola @andrewlee94 @Robbybp Thank you for the comments. I think I addressed all of them. Using the IDAES logger and ConfigDict are both great suggestions and on the TODO list. #204

blnicho commented on setup.py:
```diff
@@ -68,6 +68,7 @@ def rglob(path, glob):
         "sympy",
         "tinydb",
         "rbfopt",
+        "scipy",
```
One minor change, you should move "scipy" up in the list to keep the requirements in alphabetical order.

blnicho reviewed (commented):
I have one minor comment to address before approving

Robbybp commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,616 @@
+            # Determine the number of singular values to compute
+            # The "-1" is needed to avoid an error with svds
+            n_sv = min(n_smallest_sv, min(self.n_eq, self.n_var) - 1)
+            print("Computing the",n_sv,"smallest singular value(s)")
```
PEP8: spaces between function arguments.

Robbybp commented on idaes/core/util/model_diagnostics.py:
```diff
@@ -0,0 +1,616 @@
+
+            n = len(self.s)
+        
+            print("Smallest singular value(s):")
+            for i in range(n):
+                print("%.3E" % self.s[i])
+                if self.s[i] < tol:
+                    counter += 1
```
Out of curiosity, is there any reason you check the singular values themselves, and not the ratios of these singular values to the max singular value?

adowling2 replied:
I have not given it too much thought. Perhaps using absolute singular values tells you more about the current model formulations. I wonder if scaling by the max singular value would hide scaling issues. The counterpoint is that the condition number is the ratio of the largest and smallest singular values.

adowling2 commented:
@Robbybp @blnicho Ready for you to rereview

Robbybp reviewed (commented):
There's an extraneous argument in a docstring, a print statement that could be misleading, and `find_irreducible_degenerate_sets` doesn't fail gracefully if no such set exists.  I can approve after those are addressed. 

As an aside, do you have a definition for "candidate equation" and "irreducible degenerate set?" I still don't fully understand the algorithm and think these definitions would help. It seems this approach is great for finding "numeric" degeneracies. Is it recommended for "structural" degeneracies as well?

adowling2 commented:
@Robbybp Thank you for the great feedback. The algorithm is described here: http://sites.google.wisc.edu/alex-dowling/uploads/ESCAPE25_Dowling.pdf?attredirects=0

The main idea is to solve a MILP or perform SVD to identify candidate equations. Then loop over each candidate equation and solve another MILP to construct the smallest set of equations that includes that candidate that is degenerate. Lots of room for improvements. Over the summer, I plan to add a summary of the algorithm to the IDAES docs page.
